src/ast/passes/codegen_llvm.cpp

```diff
@@ -2126,17 +2125,29 @@ void CodegenLLVM::visit(FieldAccess &acc)
     // (bitfields and _data_loc)
     if (field.type.IsIntTy() &&
         (field.bitfield.has_value() || field.is_data_loc)) {
-      Value *src = b_.CreateAdd(expr_, b_.getInt64(field.offset));
-
       if (field.bitfield.has_value()) {
         Value *raw;
         auto field_type = b_.GetType(field.type);
-        if (type.IsCtxAccess())
-          raw = b_.CreateLoad(field_type,
-                              b_.CreateIntToPtr(src,
-                                                field_type->getPointerTo()),
-                              true);
-        else {
+        if (type.IsCtxAccess()) {
+          // We preserve pointer semantics. See `probereadDatastructElem` where
+          // the logic is essentially constrained in the same way.
+          if (expr_->getType() != b_.GET_PTR_TY()) {
+            if (!expr_->getType()->isPointerTy())
```

Comment:
I'm confused here - what's the difference between this if check and the one above `if (expr_->getType() != b_.GET_PTR_TY()) {`

Reply:
If the type is not the correct pointer type, then it is either a) an integer, or b) a pointer of another type. Coercing into the correct pointer type is one of two different operations, depending on which is the case.

Reply:
Does the pointer being the wrong type only apply for LLVM < 15? For newer LLVM versions there's only one pointer type, right?

If that's right, could you wrap it in a `#if LLVM_VERSION_MAJOR <= 14` check please? That'll make it easier to find and remove when we eventually deprecate the older LLVM versions.

Reply:
I've refactored this change a bit (might need to iterate to ensure it's working well for all versions), which I think makes this a lot easier to understand and centralizes all these conversions. It keeps the LLVM14-specific stuff, but they are generally less scattered anyways.